src/quartz/simulator/schedule.cpp

```diff
@@ -65,68 +65,119 @@ size_t Schedule::num_down_sets() {
+
+  // Greedily merge the small fusion kernels.
+  while (true) {
+    bool has_any_remaining_kernel = false;
+    std::vector<int> current_kernel;
+    int current_kernel_size = 0;
+    for (int i = optimal_fusion_kernel_size - 1; i >= 1; i--) {
+      while (!fusion_kernels_of_size[i].empty()) {
+        if (current_kernel_size + i <= optimal_fusion_kernel_size) {
+          // Add a kernel of size i to the current kernel.
+          if (result_kernels != nullptr) {
+            current_kernel.insert(
+                current_kernel.end(),
+                kernels[fusion_kernels_of_size[i].back()].first.begin(),
+                kernels[fusion_kernels_of_size[i].back()].first.end());
+          }
+          current_kernel_size += i;
+          fusion_kernels_of_size[i].pop_back();
+        } else {
+          has_any_remaining_kernel = true;
+          break;
+        }
+      }
+    }
+    if (current_kernel_size != 0) {
+      result_cost += kernel_costs[current_kernel_size];
       if (result_kernels != nullptr) {
-        result_kernels->emplace_back(kernels[index]);
+        std::sort(current_kernel.begin(), current_kernel.end());
+        result_kernels->emplace_back(std::move(current_kernel),
+                                     KernelType::fusion);
       }
-      result_cost += kernel_costs[i];
+    }
+    if (!has_any_remaining_kernel) {
+      break;
     }
   }
-  // Greedily select the small kernels.
+
+  // Greedily merge the small shared-memory kernels.
   while (true) {
     bool has_any_remaining_kernel = false;
     std::vector<int> current_kernel;
     int current_kernel_size = 0;
-    for (int i = optimal_kernel_size - 1; i >= 1; i--) {
-      while (!kernels_of_size[i].empty()) {
-        if (current_kernel_size + i <= optimal_kernel_size) {
+    for (int i = max_shared_memory_kernel_size - 1; i >= 1; i--) {
+      while (!shared_memory_kernels_of_size[i].empty()) {
+        if (current_kernel_size + i <= max_shared_memory_kernel_size) {
           // Add a kernel of size i to the current kernel.
           if (result_kernels != nullptr) {
-            current_kernel.insert(current_kernel.end(),
-                                  kernels[kernels_of_size[i].back()].begin(),
-                                  kernels[kernels_of_size[i].back()].end());
+            current_kernel.insert(
+                current_kernel.end(),
+                kernels[shared_memory_kernels_of_size[i].back()].first.begin(),
+                kernels[shared_memory_kernels_of_size[i].back()].first.end());
           }
           current_kernel_size += i;
-          kernels_of_size[i].pop_back();
+          shared_memory_kernels_of_size[i].pop_back();
         } else {
           has_any_remaining_kernel = true;
           break;
         }
       }
     }
     if (current_kernel_size != 0) {
-      result_cost += kernel_costs[current_kernel_size];
+      result_cost += kernel_cost.get_shared_memory_init_cost();
```

Comment:
Not sure if I understand correctly: if merged, seems that we should minus a init cost?